framework/src/utils/MortarUtils.C

```diff
@@ -127,6 +159,35 @@ projectQPoints3d(const Elem * const msm_elem,
     }
   };
 
+  auto sub_element_type = [primal_type, sub_elem]()
+  {
+    switch (primal_type)
+    {
+      case TRI3:
+      case QUAD4:
+      case TRI6:
+      case QUAD9:
+        return primal_type;
+      case QUAD8:
+        switch (sub_elem)
+        {
+          case 0:
+          case 1:
+          case 2:
+          case 3:
+            return TRI6;
```

Comment:
Why does this line return `TRI3` instead of `TRI6`?

Reply:
I'm also confused here, it looks like you're setting up the sub_elem_type for down below but as far as I remember we just use the primal shape functions for that? The quadrature points are on the mortar segments (x0), so we use the segments to express those but this projection is trying to find a point (x1) on the primary interface; there aren't any segments on the primary side and we just use the original shape functions.

I could be misremembering though, it's been a while!

Reply:
Sorry I'm starting to remember, primal elem isn't just the primary side, this is projecting points from the mortar segment mesh back onto both primary and secondary sides. This is just mapping the point from parameter space (reference coordinates) on mortar segment mesh elements to parameter space (reference coordinates) on the interfaces. If you use a different element type here then the quadrature point won't be in the correct parameter space for the element computations to use later.

Reply:
Wait a second, you're right, you should be using subelement types, then the quadrature point is mapped back with transform_qp, my bad! But the subelement type should correspond to the linearized elements just as before. So Quad9 should have a bunch of Quad4 subelements, TRI6 should have TRI3 subelements, Quad8 should have four triangle subelements and 1 Quad4 subelement.

Reply:
Sorry for all the confusion, it's all coming back to me now! 

This function is supposed to project from mortar segments to _linearized_ primal elements, there's a figure depicting the linearized elements (subelements) on the [documentation](https://mooseframework.inl.gov/source/constraints/AutomaticMortarGeneration.html) (I don't know why the figure is so big, I'm a markdown noob and didn't know what I was doing). After we've projected onto the linear subelements, transform_qp maps from the linearized subelements back to the original second order element (if the original element is first order we don't need to do anything else). So this projection routine should only be called on the linearized subelements and we shouldn't need any cases for second order elements (In fact we should probably verify that it's only being called on first order elements).

Reply:
Ok, I finally see what's going on!😅 

This looks correct but I agree with Antonio that it would be much more clear to put the first order element types instead of the second order ones. Below we're still calling the element type with FIRST, so it's still giving the correct linear shape functions and the point x1 is being computed correctly but it's a bit obscure (I did the same thing, oops). 

Sorry again for all the confusion, it takes a while to remember!